File: 18/18.py
import sys


def neighbors(pos):
    return (
        (pos[0] + 1, pos[1], pos[2]),
        (pos[0] - 1, pos[1], pos[2]),
        (pos[0], pos[1] + 1, pos[2]),
        (pos[0], pos[1] - 1, pos[2]),
        (pos[0], pos[1], pos[2] + 1),
        (pos[0], pos[1], pos[2] - 1),
    )


def manhatten_distance(pos1, pos2):
    return abs(pos1[0] - pos2[0]) + abs(pos1[1] - pos2[1]) + abs(pos1[2] - pos2[2])


def surface_area(cube_positions: set):
    area = 6 * len(cube_positions)

    for pos in cube_positions:
        for neighbor in neighbors(pos):
            if neighbor in cube_positions:
                area -= 1

    return area


def in_exterior(pos, cube_positions, heuristic):
    dest = (0, 0, 0)

    distances = {pos: 0}
    # in_queue = set([pos])

    prio_q = [pos]
    while prio_q:
        node = min(
            prio_q, key=lambda pos: distances[pos] + heuristic(pos, dest))
        prio_q.remove(node)

        if node == dest:
            return True

        node_neighbors = set(neighbors(node)) - cube_positions

        for neighbor in node_neighbors:
            if neighbor not in distances:
                distances[neighbor] = distances[node] + 1
                prio_q.append(neighbor)
            elif distances[node] + 1 < distances[neighbor]:
                distances[neighbor] = distances[node] + 1

    return False


def exterior_surface_area(cube_positions: set):
    area = 6 * len(cube_positions)

    for pos in cube_positions:
        for neighbor in neighbors(pos):
            if neighbor in cube_positions:
                area -= 1
            elif not in_exterior(neighbor, cube_positions, manhatten_distance):
                area -= 1
        print(pos)

    return area


def in_bounds(pos, min_xyz, max_xyz):
    return min_xyz[0] <= pos[0] <= max_xyz[0] and \
        min_xyz[1] <= pos[1] <= max_xyz[1] and \
        min_xyz[2] <= pos[2] <= max_xyz[2]


def exterior_surface_area_2(cube_positions):
    area = 0

    min_xyz = tuple(min(vals) - 1 for vals in zip(*cube_positions))
    max_xyz = tuple(max(vals) + 1 for vals in zip(*cube_positions))

    print(min_xyz, max_xyz)

    enqueued = set()
    queue = [min_xyz]

    while queue:
        pos = queue.pop(0)

        for neighbor in neighbors(pos):
            if neighbor in cube_positions:
                area += 1
            elif in_bounds(neighbor, min_xyz, max_xyz) and \
                    neighbor not in enqueued:
                queue.append(neighbor)
                enqueued.add(neighbor)

    return area


droplet = {tuple(map(int, line.strip().split(','))) for line in sys.stdin}


# print(min(droplet, key=lambda p: p[0]))
# print(min(droplet, key=lambda p: p[1]))
# print(min(droplet, key=lambda p: p[2]))

# print(max(droplet, key=lambda p: p[0]))
# print(max(droplet, key=lambda p: p[1]))
# print(max(droplet, key=lambda p: p[2]))

print(exterior_surface_area_2(droplet))

File: 18/18.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"

	"example.com/prob18/position"
)

func surfaceArea(object []position.Position) int {
	area := 6 * len(object)

	for _, pos1 := range object {
		for _, pos2 := range object {
			if position.ManhattenDistance(pos1, pos2) == 1 {
				area -= 1
			}
		}
	}

	return area
}

func minMaxXYZ(object []position.Position) (minPos, maxPos position.Position) {
	if len(object) == 0 {
		return minPos, maxPos
	}

	minPos = object[0]
	maxPos = object[0]

	for _, pos := range object {
		if pos.X < minPos.X {
			minPos.X = pos.X
		}
		if pos.Y < minPos.Y {
			minPos.Y = pos.Y
		}
		if pos.Z < minPos.Z {
			minPos.Z = pos.Z
		}

		if pos.X > maxPos.X {
			maxPos.X = pos.X
		}
		if pos.Y > maxPos.Y {
			maxPos.Y = pos.Y
		}
		if pos.Z > maxPos.Z {
			maxPos.Z = pos.Z
		}
	}

	return minPos, maxPos
}

func exteriorSurfaceArea(object []position.Position) int {
	minPos, maxPos := minMaxXYZ(object)
	minPos.X, minPos.Y, minPos.Z = minPos.X-1, minPos.Y-1, minPos.Z-1
	maxPos.X, maxPos.Y, maxPos.Z = maxPos.X+1, maxPos.Y+1, maxPos.Z+1

	fmt.Println(minPos, maxPos)

	grid := map[position.Position]bool{}
	enqueued := map[position.Position]bool{}

	for _, pos := range object {
		grid[pos] = true
	}

	area := 0

	start := minPos
	queue := []position.Position{start}

	enqueued[start] = true

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		for _, neighbor := range node.Neighbors() {
			if neighbor.InRange(minPos, maxPos) {
				_, inObject := grid[neighbor]
				_, wasEnqueued := enqueued[neighbor]

				if inObject {
					area += 1
				} else if !wasEnqueued {
					queue = append(queue, neighbor)
					enqueued[neighbor] = true
				}
			}
		}

		// fmt.Println(queue)
	}

	// for x := 0; x < MaxSize; x += 1 {
	// 	// fmt.Print("\033[H\033[2J")

	// 	for y := 0; y < MaxSize; y += 1 {
	// 		for z := 0; z < MaxSize; z += 1 {
	// 			if grid[x][y][z] {
	// 				fmt.Print("#")
	// 			} else if visited[x][y][z] {
	// 				fmt.Print(".")
	// 			} else {
	// 				fmt.Print(" ")
	// 			}
	// 		}
	// 		fmt.Println()
	// 	}

	// 	fmt.Printf("x = %d\n", x)

	// 	// time.Sleep(time.Second * 2)
	// }

	return area
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	droplet := []position.Position{}

	for scanner.Scan() {
		line := scanner.Text()
		line = strings.Trim(line, " \t\n")

		pos_strs := strings.Split(line, ",")

		pos := position.FromStringSlice(pos_strs)

		droplet = append(droplet, pos)
	}

	fmt.Printf("%d\n", exteriorSurfaceArea(droplet))
}
